controllers/batchscheduler/volcano/volcano.go

```diff
@@ -0,0 +1,193 @@
+/*
+Copyright 2020 Google LLC.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by statelicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package volcano
+
+import (
+	"fmt"
+
+	"golang.org/x/net/context"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	scheduling "volcano.sh/volcano/pkg/apis/scheduling/v1beta1"
+	volcanoclient "volcano.sh/volcano/pkg/client/clientset/versioned"
+
+	"github.com/googlecloudplatform/flink-operator/api/v1beta1"
+	schedulerinterface "github.com/googlecloudplatform/flink-operator/controllers/batchscheduler/interface"
+	"github.com/googlecloudplatform/flink-operator/controllers/model"
+)
+
+const (
+	schedulerName = "volcano"
+)
+
+// volcano scheduler implements the BatchScheduler interface.
+type VolcanoBatchScheduler struct {
+	volcanoClient volcanoclient.Interface
+}
+
+// Create volcano BatchScheduler
+func New() (schedulerinterface.BatchScheduler, error) {
+	config, err := clientcmd.BuildConfigFromFlags("", "")
+	if err != nil {
+		return nil, err
+	}
+
+	vcClient, err := volcanoclient.NewForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize volcano client with error %v", err)
+	}
+	return &VolcanoBatchScheduler{
+		volcanoClient: vcClient,
+	}, nil
+}
+
+// Name returns the current scheduler name.
+func (v *VolcanoBatchScheduler) Name() string {
+	return schedulerName
+}
+
+// Schedule reconciles batch scheduling
+func (v *VolcanoBatchScheduler) Schedule(cluster *v1beta1.FlinkCluster, state *model.DesiredClusterState) error {
+	res, size := getClusterResource(state)
```

Comment:
I feel we should move this function out of this file to keep this class Flink agnostic, Schedule() then takes Resources as the parameter. In this way, we will have a cleaner interface.

Reply:
To be more extensive,  i think pass the state is more generic. We are not sure what the batch scheduler really does from the higher level